lib/name.mli

```diff
@@ -21,43 +23,62 @@
     @author Richard Mortier <mort\@cantab.net> (documentation)
 *)
 
-open Cstruct
-
 (** DNS label, including pointer and zero. *)
 type label
 
-(** Domain name, as a list of strings. *)
-type domain_name = string list
+(** Domain name, as a list of labels ordered from the leaf to the root. *)
+type t
+
+(** Lookup key for the {! Trie}. *)
+type key = string
 
 (** Domain name map *)
-module Map: Map.S with type key = domain_name
+module Map: Map.S with type key = t
+
+(** [empty] is the empty {! t}. *)
+val empty: t
+
+(** [to_string_list name] is the label list corresponding to [name]. *)
+val to_string_list : t -> string list
+
+(** [of_string_list slist] is the domain name corresponding to label
+    list [slist]. *)
+val of_string_list : string list -> t
 
-(** Render a {! domain_name} to a simple string. *)
-val domain_name_to_string : domain_name -> string
+(** [append a b] is the domain name of the concatenation of [a] and [b]. *)
+val append: t -> t -> t
 
-(** Convert a standard domain {! string} to a {! domain_name}. *)
-val string_to_domain_name : string -> domain_name
+(** [cons label name] is the domain name with subdomain [label] under [name]. *)
+val cons: string -> t -> t
 
-(** Construct name for reverse lookup given an IPv4 address. *)
-val for_reverse : Ipaddr.V4.t -> domain_name
+(** [to_string name] is the normal, human-readable string
+    serialization of [name] without the trailing dot. *)
+val to_string : t -> string
 
-(** Parse a {! domain_name} out of a {! Cstruct.t} given a set of already
+(** [of_string name] is the domain name parsed out of the normal,
+    human-readable serialization [name]. *)
+val of_string : string -> t
+
+(** [for_reverse ip] is the name used for reverse lookup of IP address [ip]. *)
+val for_reverse : Ipaddr.t -> t
```

Comment:
why not just call this `of_ipaddr`?  The `for_reverse` is a bit confusing until you examine the type more closely.